File: Grbl-Panel/GrblMacroButtons.vb
Imports System.Windows.Forms
Imports GrblPanel.My

Partial Class GrblGui
    Private bDataChanged As Boolean = False
    ''' <summary>
    ''' The fixed names of Macros in Settings
    ''' </summary>
    Private _macroNames() As String = {"Macro1", "Macro2", "Macro3", "Macro4", "Macro5"}

    Private Sub GrblMacroButtons_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        Dim retval As Integer
        If bDataChanged Then
            retval = MsgBox(Resources.GrblGui_GrblMacroButtons_FormClosing_AreYouSureYouWantToExitWithoutSavingYourMacroChanges, MsgBoxStyle.YesNo + MsgBoxStyle.Critical, Resources.GrblGui_GrblMacroButtons_FormClosing_ConfirmExitWithoutSaving)
            If retval = vbYes Then
                e.Cancel = False
            Else
                e.Cancel = True
            End If
        End If
    End Sub

    Private Sub GrblMacroButtons_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        FillListFromSettings()
        With dgMacros
            .Columns(0).Width = .Width         ' resize the first column to the width of the entire grid, this hides the Gcode column
            'If .RowCount = 0 Then              ' if there are no macros add a default one to the list as an example
            '    .Rows.Add("Probe", "G38.2 Z-30 F10")
            '    bDataChanged = True
            'End If
        End With
        btnAdd.Text = Resources.GrblGui_GrblMacroButtons_Load_Add
        tbName.Text = ""
        tbGCode.Text = ""
    End Sub

    Private Sub FillListFromSettings()
        Dim sTemp() As String

        With dgMacros
            .RowCount = 0                                    ' make sure the grid is empty
            For Each KeyName As String In _macroNames ' add records to the grid 
                sTemp = Split(My.Settings(KeyName), ":")
                If sTemp.Count = 2 Then
                    .Rows.Add(sTemp(0), sTemp(1))
                End If
            Next
            bDataChanged = False                             ' reset our changed data flag
        End With

    End Sub

    Private Sub btnOK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOK.Click
        SaveMacros()
        Me.DialogResult = System.Windows.Forms.DialogResult.OK
    End Sub

    Private Sub SaveMacros()

        Dim index As Integer = 0

        For Each macro In _macroNames
            My.Settings(macro) = ""
        Next

        For Each row As DataGridViewRow In dgMacros.Rows
            My.Settings(_macroNames(index)) = row.Cells(0).Value.ToString + ":" + row.Cells(1).Value.ToString
            index += 1
        Next
        My.Settings.Save()
        bDataChanged = False

        ' reload onto MDI display
        EnableMacroButtons()
    End Sub

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        Me.DialogResult = System.Windows.Forms.DialogResult.Cancel
    End Sub

    Private Sub dgMacros_CellValueChanged(sender As Object, e As DataGridViewCellEventArgs) Handles dgMacros.CellValueChanged
        bDataChanged = True
    End Sub

    Private Sub dgMacros_DoubleClick(sender As Object, e As EventArgs) Handles dgMacros.DoubleClick
        With dgMacros
            If .CurrentRow.Index >= 0 Then
                tbName.Text = .Item(0, .CurrentRow.Index).Value
                tbGCode.Text = .Item(1, .CurrentRow.Index).Value
                btnAdd.Text = Resources.GrblGui_dgMacros_DoubleClick_Update
            End If
        End With
    End Sub

    Private Sub btnDeleteMacro_Click(sender As Object, e As EventArgs) Handles btnDeleteMacro.Click
        Dim retval As Integer
        Dim sMsg As String

        With dgMacros
            If .CurrentRow.Index >= 0 Then
                sMsg = Resources.GrblGui_btnDeleteMacro_Click_AreYouSureYouWantToDeleteThe & .Item(0, .CurrentRow.Index).Value.ToString & " macro?"
                retval = MsgBox(sMsg, MsgBoxStyle.YesNo + MsgBoxStyle.Critical, Resources.GrblGui_btnDeleteMacro_Click_ConfirmDelete)
                If retval = vbYes Then
                    .Rows.Remove(.CurrentRow)
                    bDataChanged = True
                End If
            End If
        End With

    End Sub

    Private Sub UpdateToolTip(sender As Object, e As EventArgs) Handles btnOK.MouseHover, btnCancel.MouseHover,
                                tbName.MouseHover, btnDeleteMacro.MouseHover, tbGCode.MouseHover,
                                dgMacros.MouseHover

        Dim sTemp As String

        If sender Is tbName Then
            sTemp = Resources.GrblGui_UpdateToolTip_NameAppearsOnTheButtonSoKeepItSmall
        ElseIf sender Is btnDeleteMacro Then
            sTemp = Resources.GrblGui_UpdateToolTip_DeleteTheSelectedMacro
        ElseIf sender Is dgMacros Then
            sTemp = Resources.GrblGui_UpdateToolTip_DblClickNameToEditMacro
        ElseIf sender Is btnCancel Then
            sTemp = Resources.GrblGui_UpdateToolTip_GetMeOuttaHereCancelAllChanges
        ElseIf sender Is tbGCode Then
            sTemp = Resources.GrblGui_UpdateToolTip_GCodeToSendWhenTheButtonIsClicked
        ElseIf sender Is btnOK Then
            sTemp = Resources.GrblGui_UpdateToolTip_CommitAllChangesToSettingsAndClose
        Else
            sTemp = Resources.GrblGui_UpdateToolTip_LimitIs5Macros
        End If

        lblStatusLabel.Text = sTemp
    End Sub

    Private Sub btnAdd_Click(sender As Object, e As EventArgs) Handles btnAdd.Click
        Dim bMatchFound As Boolean

        If tbName.Text <> "" Then
            If tbGCode.Text <> "" Then
                If btnAdd.Text = Resources.GrblGui_dgMacros_DoubleClick_Update Then
                    For Each row As DataGridViewRow In dgMacros.Rows
                        If row.Cells(0).Value.ToString = tbName.Text Then
                            row.Cells(1).Value = tbGCode.Text
                            bMatchFound = True
                            Exit For
                        End If
                    Next
                    ' if the user changed the name we cannot update so we need to add
                    If Not bMatchFound Then
                        dgMacros.Rows.Add(tbName.Text, tbGCode.Text)
                    End If
                Else
                    dgMacros.Rows.Add(tbName.Text, tbGCode.Text)
                End If
                bDataChanged = True
                btnAdd.Text = Resources.GrblGui_GrblMacroButtons_Load_Add
                tbGCode.Text = ""
                tbName.Text = ""
            Else
                MsgBox(Resources.GrblGui_btnAdd_Click_YouNeedToAddSomeGCodeToSaveAMacro, MsgBoxStyle.Information, Resources.GrblGui_btnAdd_Click_DataValidationError)
            End If
        Else
            MsgBox(Resources.GrblGui_btnAdd_Click_YouCannotCreateAMacroWithoutAName, MsgBoxStyle.Information, Resources.GrblGui_btnAdd_Click_DataValidationError)
        End If
    End Sub

    Public ReadOnly Property macroNames As String()
        Get
            Return _macroNames
        End Get
    End Property
End Class

File: Grbl-Panel.ConvertedToC#/GrblMacroButtons.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Xml.Linq;
namespace GrblPanel
{
	partial class GrblGui
	{
		private bool bDataChanged = false;
		/// <summary>
		/// The fixed names of Macros in Settings
		/// </summary>
		private string[] _macroNames = {
			"Macro1",
			"Macro2",
			"Macro3",
			"Macro4",
			"Macro5"

		};
		private void GrblMacroButtons_FormClosing(object sender, FormClosingEventArgs e)
		{
			int retval = 0;
			if (bDataChanged) {
				retval = Interaction.MsgBox("Are you sure you want to exit without saving your Macro changes?", MsgBoxStyle.YesNo + MsgBoxStyle.Critical, "Confirm exit without saving.");
				if (retval == Constants.vbYes) {
					e.Cancel = false;
				} else {
					e.Cancel = true;
				}
			}
		}

		private void GrblMacroButtons_Load(object sender, EventArgs e)
		{
			FillListFromSettings();
			var _with1 = dgMacros;
			_with1.Columns(0).Width = _with1.Width;
			// resize the first column to the width of the entire grid, this hides the Gcode column
			//If .RowCount = 0 Then              ' if there are no macros add a default one to the list as an example
			//    .Rows.Add("Probe", "G38.2 Z-30 F10")
			//    bDataChanged = True
			//End If
			btnAdd.Text = "Add";
			tbName.Text = "";
			tbGCode.Text = "";
		}

		private void FillListFromSettings()
		{
			string[] sTemp = null;

			var _with2 = dgMacros;
			_with2.RowCount = 0;
			// make sure the grid is empty
			// add records to the grid 
			foreach (string KeyName in _macroNames) {
				sTemp = Strings.Split(GrblPanel.My.Settings(KeyName), ":");
				if (sTemp.Count() == 2) {
					_with2.Rows.Add(sTemp[0], sTemp[1]);
				}
			}
			bDataChanged = false;
			// reset our changed data flag

		}

		private void btnOK_Click(System.Object sender, System.EventArgs e)
		{
			SaveMacros();
			this.DialogResult = System.Windows.Forms.DialogResult.OK;
		}


		private void SaveMacros()
		{
			int index = 0;

			foreach (string macro_loopVariable in _macroNames) {
				macro = macro_loopVariable;
				GrblPanel.My.Settings(macro) = "";
			}

			foreach (DataGridViewRow row in dgMacros.Rows) {
				GrblPanel.My.Settings(_macroNames[index]) = row.Cells[0].Value.ToString() + ":" + row.Cells[1].Value.ToString();
				index += 1;
			}
			GrblPanel.My.Settings.Save();
			bDataChanged = false;

			// reload onto MDI display
			EnableMacroButtons();
		}

		private void btnCancel_Click(System.Object sender, System.EventArgs e)
		{
			this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
		}

		private void dgMacros_CellValueChanged(object sender, DataGridViewCellEventArgs e)
		{
			bDataChanged = true;
		}

		private void dgMacros_DoubleClick(object sender, EventArgs e)
		{
			var _with3 = dgMacros;
			if (_with3.CurrentRow.Index >= 0) {
				tbName.Text = _with3.Item(0, _with3.CurrentRow.Index).Value;
				tbGCode.Text = _with3.Item(1, _with3.CurrentRow.Index).Value;
				btnAdd.Text = "Update";
			}
		}

		private void btnDeleteMacro_Click(object sender, EventArgs e)
		{
			int retval = 0;
			string sMsg = null;

			var _with4 = dgMacros;
			if (_with4.CurrentRow.Index >= 0) {
				sMsg = "Are you sure you want to delete the " + _with4.Item(0, _with4.CurrentRow.Index).Value.ToString + " macro?";
				retval = Interaction.MsgBox(sMsg, MsgBoxStyle.YesNo + MsgBoxStyle.Critical, "Confirm Delete");
				if (retval == Constants.vbYes) {
					_with4.Rows.Remove(_with4.CurrentRow);
					bDataChanged = true;
				}
			}

		}


		private void UpdateToolTip(object sender, EventArgs e)
		{
			string sTemp = null;

			if (object.ReferenceEquals(sender, tbName)) {
				sTemp = "Name appears on the button, so keep it small";
			} else if (object.ReferenceEquals(sender, btnDeleteMacro)) {
				sTemp = "Delete the selected macro";
			} else if (object.ReferenceEquals(sender, dgMacros)) {
				sTemp = "DblClick name to edit macro";
			} else if (object.ReferenceEquals(sender, btnCancel)) {
				sTemp = "Get me outta here, cancel all changes";
			} else if (object.ReferenceEquals(sender, tbGCode)) {
				sTemp = "G-Code to send when the button is clicked.";
			} else if (object.ReferenceEquals(sender, btnOK)) {
				sTemp = "Commit all changes to Settings and close";
			} else {
				sTemp = "Limit is 5 macros";
			}

			lblStatusLabel.Text = sTemp;
		}

		private void btnAdd_Click(object sender, EventArgs e)
		{
			bool bMatchFound = false;

			if (!string.IsNullOrEmpty(tbName.Text)) {
				if (!string.IsNullOrEmpty(tbGCode.Text)) {
					if (btnAdd.Text == "Update") {
						foreach (DataGridViewRow row in dgMacros.Rows) {
							if (row.Cells[0].Value.ToString() == tbName.Text) {
								row.Cells[1].Value = tbGCode.Text;
								bMatchFound = true;
								break; // TODO: might not be correct. Was : Exit For
							}
						}
						// if the user changed the name we cannot update so we need to add
						if (!bMatchFound) {
							dgMacros.Rows.Add(tbName.Text, tbGCode.Text);
						}
					} else {
						dgMacros.Rows.Add(tbName.Text, tbGCode.Text);
					}
					bDataChanged = true;
					btnAdd.Text = "Add";
					tbGCode.Text = "";
					tbName.Text = "";
				} else {
					Interaction.MsgBox("You need to add some G-code to save a macro", MsgBoxStyle.Information, "Data Validation Error");
				}
			} else {
				Interaction.MsgBox("You cannot create a macro without a name.", MsgBoxStyle.Information, "Data Validation Error");
			}
		}

		public string[] macroNames {
			get { return _macroNames; }
		}
	}
}
